search/includes/classes/class-health.php

```diff
@@ -141,6 +144,215 @@ public static function validate_index_posts_count() {
+				$next_batch_post_id = $last_post_id + 1;
+			}
+
+			if ( $is_cli ) {
+				\WP_CLI::line( sprintf( 'Validating posts %d - %d', $start_post_id, $next_batch_post_id - 1 ) );
+			}
+			
+			$result = self::validate_index_posts_content_batch( $indexable, $start_post_id, $next_batch_post_id );
+
+			if ( is_wp_error( $result ) ) {
+				$result = [
+					'entity' => $indexable->slug,
+					'start_post_id' => $start_post_id,
+					'error' => $result->get_error_message()
+				];
+			}
+
+			$results = array_merge( $results, $result );
+
+			// Limit $results size
+			if ( strlen( serialize( $results ) ) > self::CONTENT_VALIDATION_MAX_DIFF_SIZE ) {
+				$error = new WP_Error( 'diff-size-limit-reached', sprintf( 'Reached diff size limit of %d bytes, aborting', self::CONTENT_VALIDATION_MAX_DIFF_SIZE ) );
+
+				$error->add_data( $results, 'diff' );
+
+				return $error;
+			}
+
+			$start_post_id += self::CONTENT_VALIDATION_BATCH_SIZE;
+
+			if ( $dynamic_last_post_id ) {
+				// Requery for the last post id after each batch b/c the site is probably growing
+				// while this runs
+				$last_post_id = self::get_last_post_id();
+			}
+		} while ( $start_post_id <= $last_post_id );
+
+		return $results;
+	}
+
+	public static function validate_index_posts_content_batch( $indexable, $start_post_id, $next_batch_post_id ) {
+		global $wpdb;
+
+		$sql = $wpdb->prepare( "SELECT ID, post_type, post_status FROM $wpdb->posts WHERE ID >= %d AND ID < %d", $start_post_id, $next_batch_post_id );
+
+		$rows = $wpdb->get_results( $sql );
+
+		$post_types = $indexable->get_indexable_post_types();
+		$post_statuses = $indexable->get_indexable_post_status();
+
+		// First we need to see identify which posts are actually expected in the index, by checking the same filters that
+		// are used in ElasticPress\Indexable\Post\SyncManager::action_sync_on_update()
+		$expected_post_rows = array_filter( $rows, function( $row ) use ( $post_types, $post_statuses ) {
+			
+			if ( ! in_array( $row->post_type, $post_types, true ) ) {
+				return false;
+			}
+			
+			if ( ! in_array( $row->post_status, $post_statuses, true ) ) {
+				return false;
+			}
+
+			$skipped = apply_filters( 'ep_post_sync_kill', false, $row->ID, $row->ID );
+
+			return ! $skipped;
+		} );
+
+		$document_ids = self::get_document_ids_for_batch( $start_post_id, $next_batch_post_id );
+
+		// Grab all of the documents from ES
+		$documents = $indexable->multi_get( $document_ids );
+
+		// Filter out any that weren't found
+		$documents = array_filter( $documents, function( $document ) {
+			return ! is_null( $document );
+		} );
+
+		$expected_post_ids = wp_list_pluck( $expected_post_rows, 'ID' );
+		$found_document_ids = wp_list_pluck( $documents, 'ID' );
+
+		$diffs = [];
+
+		// What's missing in ES?
+		$missing_from_index = array_diff( $expected_post_ids, $found_document_ids );
+
+		// If anything is missing from index, record it
+		if ( count( $missing_from_index ) ) {
```

Comment:
Should probably do an explicit value check for clarity rather than rely on 0 being implicitly falsey.

Reply:
👍 updated in 54e17c34